bangc-ops/mlu_op.h

```diff
@@ -2039,6 +2039,348 @@ mluOpLog(mluOpHandle_t handle,
+
+// Group:Carafe
+/*!
+ * @brief Destroys a CARAFE descriptor \b carafe_desc that is previously created by
+ *        the ::mluOpCreateCarafeDescriptor function.
+ *
+ * The CARAFE descriptor is defined in ::mluOpCarafeDescriptor_t
+ * and holds the information about the CARAFE forward or backward operations.
+ *
+ * @param[in] carafe_desc
+ * The CARAFE descriptor to be destroyed. For detailed information,
+ * see ::mluOpCarafeDescriptor_t.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @note
+ * - You need to call this function after calling the ::mluOpCarafeForward,
+ *   or ::mluOpCarafeBackward function. Otherwise, \p MLUOP_STATUS_BAD_PARAM is returned.
+ * - This function should be called to destroy the CARAFE descriptor.
+ *   Otherwise, memory leak may occur.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - None.
+ */
+mluOpStatus_t MLUOP_WIN_API
+mluOpDestroyCarafeDescriptor(mluOpCarafeDescriptor_t carafe_desc);
+
+// Group:Carafe
+/*!
+ * @brief Performs the CARAFE (Content-Aware ReAssembly of FEatures) upsampling operation
+ *        on the input feature maps \b input using weighted combination, where the
+ *        filter are set with \b mask. The upsampled feature maps are returned in
+ *        the output tensor \b output.
+ *
+ * CARAFE performs upsampling at each output location by weighted summation in a nearby mask
+ * window around the corresponding input location. The mask filter are defined separately
+ * for each output location, which offers the ability of content-aware handling.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and
+ * queues in the carafe forward operation. For detailed information,
+ * see ::mluOpHandle_t.
+ * @param[in] carafe_desc
+ * The descriptor of the CARAFE operator. For detailed information,
+ * see ::mluOpCarafeDescriptor_t.
+ * @param[in] input_desc
+ * The tensor descriptor of the input feature maps,
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] input
+ * Pointer to the MLU memory that stores the input tensor.
+ * @param[in] mask_desc
+ * The tensor descriptor of the mask applied to the input feature maps,
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] mask
+ * Pointer to the MLU memory that stores the mask tensor.
+ * @param[in] output_desc
+ * The tensor descriptor of the output upsampled feature maps,
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[out] output
+ * Pointer to the MLU memory that stores the output tensor.
+ * 
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_NOT_SUPPORTED
+ *
+ * @par Formula
+ * - See "CARAFE Operator" section in "Cambricon MLUOP User Guide" for details
+ *
+ * @par Data Type
+ * - \b input: half, float.
+ * - \b mask: half, float.
+ * - \b output: half, float.
+ * - Data types of \b input, \b mask and \b output tensors must be the same
```

Comment:
@ par Data Type
- Data types of \b input, \b mask and \b output tensors must be the same。
- The supported data types of input， output, and mask tensors are as follows:
   - input: half, float
   - mask: half, float
   - output: half, float